### PR TITLE
Add number representation assists

### PR DESCRIPTION
**Pull request by errx**
Reimplemented assists from this PR https://github.com/rust-analyzer/rust-analyzer/pull/3683 with current APIs.
image
image

I've decided not to add options about size of the groups so behaviour is similar to clippy's. 
Minimal number length is also taken from clippy.

### CODE REVIEW

File: crates/ide_assists/src/handlers/number_representation.rs [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    let range = literal.syntax().text_range();
+    acc.add_group(
+        &group_id,
+        AssistId("reformat_number_literal", AssistKind::RefactorInline),
+        label,
+        range,
+        |builder| builder.replace(range, converted),
+    );
+
+    Some(())
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/number_representation.rs:53]:
```suggestion
    acc.add_group(
        &group_id,
        AssistId("reformat_number_literal", AssistKind::RefactorInline),
        label,
        range,
        |builder| builder.replace(range, converted),
    )
```

Comment (errx) [reply crates/ide_assists/src/handlers/number_representation.rs]: Thanks, fixed.

File: crates/ide_assists/src/handlers/number_representation.rs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+    let value = literal.str_value();
+    if value.len() < MIN_NUMBER_OF_DIGITS_TO_FORMAT {
+        return None;
+    }
+
+    let radix = literal.radix();
+    let mut converted = literal.prefix().to_string();
+    converted.push_str(&add_group_separators(literal.str_value(), group_size(radix)));
+    if let Some(suffix) = literal.suffix() {
```
[the remark this answers is not in the document]
Comment (errx) [reply crates/ide_assists/src/handlers/number_representation.rs:38]: I didn't want to make IntNumberParts public. Is it ok to return tuple (&str, &str, &str) from this function?

[the remark this answers is not in the document]
Comment (Veykril) [reply crates/ide_assists/src/handlers/number_representation.rs:38]: Ye that is fine

[the remark this answers is not in the document]
Comment (errx) [reply crates/ide_assists/src/handlers/number_representation.rs]: Done

**Comment by Veykril** [comment]
Thanks!
bors r+

File: crates/ide_assists/src/handlers/number_representation.rs [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+    let group_id = GroupLabel("Reformat number literal".into());
+    let label = format!("Convert {} to {}", literal, converted);
+    let range = literal.syntax().text_range();
+    acc.add_group(
+        &group_id,
+        AssistId("reformat_number_literal", AssistKind::RefactorInline),
+        label,
+        range,
+        |builder| builder.replace(range, converted),
+    )
```
Comment (iDawer) [inline crates/ide_assists/src/handlers/number_representation.rs:49]: May I nitpick little bit? :smile:  Things that related to assist application could better reside in `|builder| {... }` closure. Outside are only steps that determine applicability. I.e. application is computed lazily. Doing so reduces _light bulb_ computation complexity.

Comment (Veykril) [reply crates/ide_assists/src/handlers/number_representation.rs:49]: Oh I completely blanked with that when reviewing yes, those things should go inside the closure.

Comment (errx) [reply crates/ide_assists/src/handlers/number_representation.rs:49]: Thanks. I'll try to improve it in another mr.

Comment (errx) [reply crates/ide_assists/src/handlers/number_representation.rs:49]: @iDawer but I'am using formatted value in the label. Do you think it's better to keep label static, something like "Add digit separators"?

Comment (Veykril) [reply crates/ide_assists/src/handlers/number_representation.rs:49]: Right you are using the formatted string in there(I think the integer base conversion assist does the same). Given that this only applies on longer integer literals anyways without doing a lot of calculations either this is probably fine as is.

**Pull request merged**